src/data/tree.lean

```diff
@@ -81,4 +84,69 @@ def map {β} (f : α → β) : tree α → tree β
 | nil := nil
 | (node a l r) := node (f a) (map l) (map r)
 
+/-- A non-nil `tree α`; useful when we want an arbitrary value other than `nil` -/
+abbreviation non_nil [inhabited α] : tree α := node default nil nil
+
+@[simp] lemma non_nil_ne [inhabited α] : non_nil ≠ (@nil α) := by trivial
+
+/-- The number of internal nodes (i.e. not including leaves) of a binary tree -/
+@[simp] def nodes : tree α → ℕ
+| nil := 0
+| (node _ a b) := a.nodes + b.nodes + 1
+
+/-- The number of leaves of a binary tree -/
+@[simp] def num_leaves : tree α → ℕ
+| nil := 1
+| (node _ a b) := a.num_leaves + b.num_leaves
+
+/-- The height - length of the longest path from the root - of a binary tree -/
+@[simp] def height : tree α → ℕ
+| nil := 0
+| (node _ a b) := max a.height b.height + 1
+
+lemma leaves_eq_nodes_succ (x : tree α) : x.num_leaves = x.nodes + 1 :=
+by { induction x; simp [*, nat.add_comm, nat.add_assoc, nat.add_left_comm], }
+
+lemma leaves_pos (x : tree α) : 0 < x.num_leaves :=
+by { rw leaves_eq_nodes_succ, exact x.nodes.zero_lt_succ, }
+
+lemma height_le_nodes : ∀ (x : tree α), x.height ≤ x.nodes
+| nil := le_refl _
```

Comment:
```suggestion
| nil := le_rfl
```

Reply:
I didn't use `le_rfl` because it requires the import from `order.basic`. Not a big deal, but I thought it was silly to add an import to save two characters.

Reply:
I think you can just add the import.

Reply:
I agree that this golf doesn't justify an extra import.